File: PlayerRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using UnityEngine;

namespace BattleFleet.Database
{
    /// <summary>
    /// Player Repository - CRUD operations for Player table
    /// Handles all player-related database operations
    /// </summary>
    public class PlayerRepository
    {
        private DatabaseManager dbManager;

        public PlayerRepository()
        {
            dbManager = DatabaseManager.Instance;
        }

        /// <summary>
        /// CREATE - Insert a new player into the database
        /// </summary>
        public Player CreatePlayer(string playerName)
        {
            if (string.IsNullOrWhiteSpace(playerName))
            {
                Debug.LogError("Player name cannot be empty.");
                return null;
            }

            try
            {
                // Check if player already exists
                if (PlayerExists(playerName))
                {
                    Debug.LogWarning($"Player '{playerName}' already exists.");
                    return null;
                }

                // Insert player
                string insertPlayerQuery = @"INSERT INTO Players (PlayerName, CreatedDate, IsActive) 
                                            VALUES (@playerName, @createdDate, 1)";
                
                Dictionary<string, object> playerParams = new Dictionary<string, object>
                {
                    { "@playerName", playerName },
                    { "@createdDate", DateTime.Now }
                };

                int result = dbManager.ExecuteNonQuery(insertPlayerQuery, playerParams);
                
                if (result <= 0)
                {
                    Debug.LogError("Failed to create player.");
                    return null;
                }

                // Get the newly created player ID
                int playerID = GetPlayerIDByName(playerName);
                
                // Create associated PlayerStats record
                CreatePlayerStats(playerID);

                // Retrieve and return the created player
                Player newPlayer = GetPlayerByID(playerID);
                Debug.Log($"Player '{playerName}' created successfully with ID: {playerID}");
                
                return newPlayer;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error creating player: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// READ - Get player by ID
        /// </summary>
        public Player GetPlayerByID(int playerID)
        {
            try
            {
                string query = "SELECT * FROM Players WHERE PlayerID = @playerID";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerID", playerID }
                };

                IDataReader reader = dbManager.ExecuteReader(query, parameters);
                
                if (reader != null && reader.Read())
                {
                    Player player = new Player
                    {
                        PlayerID = Convert.ToInt32(reader["PlayerID"]),
                        PlayerName = reader["PlayerName"].ToString(),
                        CreatedDate = Convert.ToDateTime(reader["CreatedDate"]),
                        IsActive = Convert.ToInt32(reader["IsActive"]) == 1
                    };

                    if (reader["LastPlayedDate"] != DBNull.Value)
                    {
                        player.LastPlayedDate = Convert.ToDateTime(reader["LastPlayedDate"]);
                    }

                    reader.Close();
                    return player;
                }

                if (reader != null) reader.Close();
                return null;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving player by ID: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// READ - Get player by name
        /// </summary>
        public Player GetPlayerByName(string playerName)
        {
            try
            {
                string query = "SELECT * FROM Players WHERE PlayerName = @playerName";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerName", playerName }
                };

                IDataReader reader = dbManager.ExecuteReader(query, parameters);
                
                if (reader != null && reader.Read())
                {
                    Player player = new Player
                    {
                        PlayerID = Convert.ToInt32(reader["PlayerID"]),
                        PlayerName = reader["PlayerName"].ToString(),
                        CreatedDate = Convert.ToDateTime(reader["CreatedDate"]),
                        IsActive = Convert.ToInt32(reader["IsActive"]) == 1
                    };

                    if (reader["LastPlayedDate"] != DBNull.Value)
                    {
                        player.LastPlayedDate = Convert.ToDateTime(reader["LastPlayedDate"]);
                    }

                    reader.Close();
                    return player;
                }

                if (reader != null) reader.Close();
                return null;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving player by name: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// READ - Get all active players
        /// </summary>
        public List<Player> GetAllActivePlayers()
        {
            List<Player> players = new List<Player>();

            try
            {
                string query = "SELECT * FROM Players WHERE IsActive = 1 ORDER BY PlayerName ASC";
                IDataReader reader = dbManager.ExecuteReader(query);

                if (reader != null)
                {
                    while (reader.Read())
                    {
                        Player player = new Player
                        {
                            PlayerID = Convert.ToInt32(reader["PlayerID"]),
                            PlayerName = reader["PlayerName"].ToString(),
                            CreatedDate = Convert.ToDateTime(reader["CreatedDate"]),
                            IsActive = Convert.ToInt32(reader["IsActive"]) == 1
                        };

                        if (reader["LastPlayedDate"] != DBNull.Value)
                        {
                            player.LastPlayedDate = Convert.ToDateTime(reader["LastPlayedDate"]);
                        }

                        players.Add(player);
                    }
                    reader.Close();
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving all players: {e.Message}");
            }

            return players;
        }

        /// <summary>
        /// UPDATE - Update player information
        /// </summary>
        public bool UpdatePlayer(Player player)
        {
            if (player == null || player.PlayerID <= 0)
            {
                Debug.LogError("Invalid player object.");
                return false;
            }

            try
            {
                string query = @"UPDATE Players 
                                SET PlayerName = @playerName, 
                                    LastPlayedDate = @lastPlayedDate,
                                    IsActive = @isActive
                                WHERE PlayerID = @playerID";

                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerName", player.PlayerName },
                    { "@lastPlayedDate", player.LastPlayedDate ?? (object)DBNull.Value },
                    { "@isActive", player.IsActive ? 1 : 0 },
                    { "@playerID", player.PlayerID }
                };

                int result = dbManager.ExecuteNonQuery(query, parameters);
                
                if (result > 0)
                {
                    Debug.Log($"Player '{player.PlayerName}' updated successfully.");
                    return true;
                }
                
                return false;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error updating player: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// UPDATE - Update last played date for a player
        /// </summary>
        public bool UpdateLastPlayedDate(int playerID)
        {
            try
            {
                string query = "UPDATE Players SET LastPlayedDate = @lastPlayedDate WHERE PlayerID = @playerID";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@lastPlayedDate", DateTime.Now },
                    { "@playerID", playerID }
                };

                int result = dbManager.ExecuteNonQuery(query, parameters);
                return result > 0;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error updating last played date: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// DELETE - Delete a player (soft delete - sets IsActive to 0)
        /// </summary>
        public bool DeletePlayer(int playerID, bool hardDelete = false)
        {
            try
            {
                if (hardDelete)
                {
                    // Hard delete - remove from database completely
                    string query = "DELETE FROM Players WHERE PlayerID = @playerID";
                    Dictionary<string, object> parameters = new Dictionary<string, object>
                    {
                        { "@playerID", playerID }
                    };

                    int result = dbManager.ExecuteNonQuery(query, parameters);
                    
                    if (result > 0)
                    {
                        Debug.Log($"Player with ID {playerID} permanently deleted.");
                        return true;
                    }
                }
                else
                {
                    // Soft delete - just mark as inactive
                    string query = "UPDATE Players SET IsActive = 0 WHERE PlayerID = @playerID";
                    Dictionary<string, object> parameters = new Dictionary<string, object>
                    {
                        { "@playerID", playerID }
                    };

                    int result = dbManager.ExecuteNonQuery(query, parameters);
                    
                    if (result > 0)
                    {
                        Debug.Log($"Player with ID {playerID} deactivated.");
                        return true;
                    }
                }

                return false;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error deleting player: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// Helper - Check if player exists by name
        /// </summary>
        public bool PlayerExists(string playerName)
        {
            try
            {
                string query = "SELECT COUNT(*) FROM Players WHERE PlayerName = @playerName";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerName", playerName }
                };

                object result = dbManager.ExecuteScalar(query, parameters);
                return Convert.ToInt32(result) > 0;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error checking if player exists: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// Helper - Get player ID by name
        /// </summary>
        public int GetPlayerIDByName(string playerName)
        {
            try
            {
                string query = "SELECT PlayerID FROM Players WHERE PlayerName = @playerName LIMIT 1";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerName", playerName }
                };

                object result = dbManager.ExecuteScalar(query, parameters);
                
                if (result != null && result != DBNull.Value)
                {
                    return Convert.ToInt32(result);
                }

                return -1;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving player ID: {e.Message}");
                return -1;
            }
        }

        /// <summary>
        /// Helper - Get total player count
        /// </summary>
        public int GetTotalPlayerCount(bool activeOnly = true)
        {
            try
            {
                string query = activeOnly ? 
                    "SELECT COUNT(*) FROM Players WHERE IsActive = 1" :
                    "SELECT COUNT(*) FROM Players";

                object result = dbManager.ExecuteScalar(query);
                return Convert.ToInt32(result);
            }
            catch (Exception e)
            {
                Debug.LogError($"Error counting players: {e.Message}");
                return 0;
            }
        }

        /// <summary>
        /// Helper - Create associated PlayerStats when creating a new player
        /// </summary>
        private bool CreatePlayerStats(int playerID)
        {
            try
            {
                string query = @"INSERT INTO PlayerStats (PlayerID, Level, TotalExperience, TotalBattles, 
                                TotalWins, TotalLosses, TotalDraws, LastUpdated) 
                                VALUES (@playerID, 1, 0, 0, 0, 0, 0, @lastUpdated)";

                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerID", playerID },
                    { "@lastUpdated", DateTime.Now }
                };

                int result = dbManager.ExecuteNonQuery(query, parameters);
                return result > 0;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error creating player stats: {e.Message}");
                return false;
            }
        }
    }
}

File: database/player_repository.py
# player_repository.py
# CRUD operations for the Players table.
# FIX (vs Unity version): get_player_by_id / get_player_by_name now read FleetLevel and FleetXP.

import logging
from datetime import datetime, timezone
from typing import Optional

from .database_manager import DatabaseManager
from .models import Player

logger = logging.getLogger(__name__)


class PlayerRepository:

    def __init__(self, db: DatabaseManager) -> None:
        # Dependency-injected instead of using an autoload singleton,
        # which is the standard Python pattern for testability.
        self._db = db

    # ── CREATE ─────────────────────────────────────────────────────────────────

    def create_player(self, player_name: str, email: str = "", password_hash: str = "") -> Optional[Player]:
        """Insert a new player and create an associated PlayerStats row.
        Returns the new Player object, or None on failure."""
        if not player_name.strip():
            logger.error("PlayerRepository: Player name cannot be empty.")
            return None

        if self.player_exists(player_name):
            logger.warning("PlayerRepository: Player '%s' already exists.", player_name)
            return None

        now = _now()
        rows_affected = self._db.execute_non_query(
            "INSERT INTO Players (PlayerName, Email, PasswordHash, CreatedDate, IsActive, ELO) "
            "VALUES (?, ?, ?, ?, 1, 0)",
            (player_name, email, password_hash, now),
        )

        if rows_affected <= 0:
            logger.error("PlayerRepository: Failed to insert player '%s'.", player_name)
            return None

        player_id = self._db.get_last_insert_id()
        self._create_player_stats(player_id, now)

        logger.info("PlayerRepository: Created player '%s' (ID: %d).", player_name, player_id)
        return self.get_player_by_id(player_id)

    # ── READ ───────────────────────────────────────────────────────────────────

    def get_player_by_id(self, player_id: int) -> Optional[Player]:
        """Returns a Player by primary key, or None if not found."""
        rows = self._db.execute_reader(
            "SELECT * FROM Players WHERE PlayerID = ?", (player_id,)
        )
        if not rows:
            return None
        return _map_player(rows[0])

    def get_player_by_name(self, player_name: str) -> Optional[Player]:
        """Returns a Player by name, or None if not found."""
        rows = self._db.execute_reader(
            "SELECT * FROM Players WHERE PlayerName = ?", (player_name,)
        )
        if not rows:
            return None
        return _map_player(rows[0])

    def get_all_active_players(self) -> list[Player]:
        """Returns all players with IsActive = 1, sorted by name."""
        rows = self._db.execute_reader(
            "SELECT * FROM Players WHERE IsActive = 1 ORDER BY PlayerName ASC"
        )
        return [_map_player(row) for row in rows]

    # ── UPDATE ─────────────────────────────────────────────────────────────────

    def update_player(self, player: Player) -> bool:
        """Updates PlayerName, LastPlayedDate and IsActive. Returns True on success."""
        if player is None or player.player_id <= 0:
            logger.error("PlayerRepository: Invalid player object.")
            return False

        result = self._db.execute_non_query(
            "UPDATE Players SET PlayerName = ?, LastPlayedDate = ?, IsActive = ? "
            "WHERE PlayerID = ?",
            (
                player.player_name,
                player.last_played_date,
                1 if player.is_active else 0,
                player.player_id,
            ),
        )
        return result > 0

    def update_last_played_date(self, player_id: int) -> bool:
        """Stamps LastPlayedDate with the current UTC time."""
        result = self._db.execute_non_query(
            "UPDATE Players SET LastPlayedDate = ? WHERE PlayerID = ?",
            (_now(), player_id),
        )
        return result > 0

    # ── DELETE ─────────────────────────────────────────────────────────────────

    def delete_player(self, player_id: int, hard_delete: bool = False) -> bool:
        """Soft delete (default) sets IsActive = 0. Hard delete removes the row."""
        if hard_delete:
            result = self._db.execute_non_query(
                "DELETE FROM Players WHERE PlayerID = ?", (player_id,)
            )
            if result > 0:
                logger.info("PlayerRepository: Player %d permanently deleted.", player_id)
        else:
            result = self._db.execute_non_query(
                "UPDATE Players SET IsActive = 0 WHERE PlayerID = ?", (player_id,)
            )
            if result > 0:
                logger.info("PlayerRepository: Player %d deactivated.", player_id)
        return result > 0

    # ── HELPERS ────────────────────────────────────────────────────────────────

    def player_exists(self, player_name: str) -> bool:
        count = self._db.execute_scalar(
            "SELECT COUNT(*) FROM Players WHERE PlayerName = ?", (player_name,)
        )
        return int(count or 0) > 0

    def get_player_id_by_name(self, player_name: str) -> int:
        result = self._db.execute_scalar(
            "SELECT PlayerID FROM Players WHERE PlayerName = ? LIMIT 1", (player_name,)
        )
        return int(result) if result is not None else -1

    def get_total_player_count(self, active_only: bool = True) -> int:
        query = (
            "SELECT COUNT(*) FROM Players WHERE IsActive = 1"
            if active_only
            else "SELECT COUNT(*) FROM Players"
        )
        result = self._db.execute_scalar(query)
        return int(result or 0)

    # ── Private ────────────────────────────────────────────────────────────────

    def _create_player_stats(self, player_id: int, now: str) -> None:
        self._db.execute_non_query(
            """INSERT INTO PlayerStats
               (PlayerID, TotalBattles,
                TotalWins, TotalLosses, TotalDraws, LastUpdated)
               VALUES (?, 0, 0, 0, 0, ?)""",
            (player_id, now),
        )


# ── Module-level helpers ───────────────────────────────────────────────────────

def _now() -> str:
    """ISO-8601 datetime string in UTC, matching SQLite's datetime('now')."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")


def _map_player(row: dict) -> Player:
    """FIX: reads FleetLevel and FleetXP which the Unity version omitted."""
    return Player(
        player_id=int(row.get("PlayerID", 0)),
        player_name=str(row.get("PlayerName", "")),
        email=str(row.get("Email", "")),
        password_hash=str(row.get("PasswordHash", "")),
        created_date=str(row.get("CreatedDate", "")),
        last_played_date=str(row.get("LastPlayedDate", "") or ""),
        is_active=int(row.get("IsActive", 1)) == 1,
        elo=int(row.get("ELO", 0)),
    )
